src/aave-v3/SupplyVault.sol

```diff
@@ -217,36 +196,61 @@ contract SupplyVault is ISupplyVault, SupplyVaultBase {
-        uint256 supply = totalSupply();
         for (uint256 i; i < rewardTokens.length; ++i) {
             address rewardToken = rewardTokens[i];
             uint256 claimedAmount = claimedAmounts[i];
+            UserRewardsData storage userRewardsData = userRewards[rewardToken][_user];
             uint128 rewardsIndexMem;
 
-            if (supply > 0 && claimedAmount > 0) {
+            if (totalSupply() > 0 && claimedAmount > 0) {
                 rewardsIndexMem =
                     rewardsIndex[rewardToken] +
-                    claimedAmount.mulDivDown(SCALE, supply).safeCastTo128();
+                    _getUnaccruedRewardIndex(claimedAmount);
                 rewardsIndex[rewardToken] = rewardsIndexMem;
             } else rewardsIndexMem = rewardsIndex[rewardToken];
 
-            UserRewardsData storage userRewardsData = userRewards[rewardToken][_user];
-            uint256 rewardsIndexDiff;
-
-            // Safe because we always have `rewardsIndex` >= `userRewardsData.index`.
-            unchecked {
-                rewardsIndexDiff = rewardsIndexMem - userRewardsData.index;
-            }
-
-            if (rewardsIndexDiff > 0) {
-                uint128 accruedRewards = balanceOf(_user)
-                .mulDivDown(rewardsIndexDiff, SCALE)
-                .safeCastTo128();
-                userRewardsData.unclaimed += accruedRewards;
+            if (rewardsIndexMem > userRewardsData.index) {
+                uint128 accruedReward = _getUnaccruedRewardFromUnaccruedRewardIndex(
+                    _user,
+                    rewardsIndexMem - userRewardsData.index
+                );
+                userRewardsData.unclaimed += accruedReward;
                 userRewardsData.index = rewardsIndexMem;
 
-                emit Accrued(rewardToken, _user, rewardsIndexMem, accruedRewards);
+                emit Accrued(rewardToken, _user, rewardsIndexMem, accruedReward);
             }
         }
     }
+
+    function _getUnaccruedReward(
+        address _user,
+        address _rewardToken,
+        uint256 _claimableReward
+    ) internal view returns (uint128 unaccruedReward) {
+        unaccruedReward = _getUnaccruedRewardFromUnaccruedRewardIndex(
+            _user,
+            _getUnaccruedRewardIndex(_claimableReward) + // The unaccrued reward index
+                rewardsIndex[_rewardToken] -
+                userRewards[_rewardToken][_user].index // The difference between the current reward index and the user's index
+        );
+    }
+
+    function _getUnaccruedRewardFromUnaccruedRewardIndex(
+        address _user,
+        uint128 _unaccruedRewardIndex
+    ) internal view returns (uint128 unaccruedReward) {
+        unaccruedReward = balanceOf(_user).mulDivDown(_unaccruedRewardIndex, SCALE).safeCastTo128();
+    }
+
+    function _getUnaccruedRewardIndex(uint256 _claimableReward)
+        internal
+        view
+        returns (uint128 unaccruedRewardIndex)
+    {
+        uint256 supply = totalSupply();
```

Comment:
I'm in favor of passing this in the arguments

Reply:
I don't have strong opinions on this, both ways are ok to me

Reply:
To me this is not really complex to do, it's just like an helper function that does the math we feed with the required argument like in the IRM or the `RewardsManager`